sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -848,15 +856,25 @@ export class MessageReceiver extends LinkEntity {
 
   /**
    * Will reconnect the receiver link if necessary.
-   * @param {AmqpError | Error} [receiverError] The receiver error if any.
+   * @param receiverError The receiver error if any.
+   * @param connectionDidDisconnect Whether this method is called as a result of a connection disconnect.
    * @returns {Promise<void>} Promise<void>.
    */
-  async onDetached(receiverError?: AmqpError | Error): Promise<void> {
+  async onDetached(
+    receiverError?: AmqpError | Error,
+    connectionDidDisconnect?: boolean
+  ): Promise<void> {
     const connectionId = this._context.namespace.connectionId;
+    // Prevent multiple onDetached invocations from running concurrently.
+    if (this._isDetaching) {
+      log.error(
+        `[${connectionId}] Receiver '${this.name}' with address '${this.address}' has had onDetached called while already detaching.`
+      );
+      return;
```

Comment:
I'm guessing the answer is 'no' but do we feel confident enough that this is an error that we should throw an error?

Reply:
Am confused. Are you saying we should throw an error or that we shouldnt?
Right now, all we are doing is logging.

Reply:
Also confused. This is simply logging, not throwing. This will only log if a previous `onDetached` hasn't completed yet (most likely case due to a previous onDetached still trying to bring the connection back up)

Reply:
Sorry, not enough punctuation :)

What I'm asking is - is this a serious error? If it is why aren't we throwing rather than just logging it? We don't check logs anywhere in our tests so if this is happening (by accident, whatever) we won't detect it and thus won't fix it.

Reply:
To give an explicit scenario on when this happens, let's say your network connection temporarily goes down.
This will trigger the connection `disconnect` event to fire, which ends up calling `onDetached`.
Now, we determine from this error that it's a networking issue, so we attempt to retry (retry is called in onDetached).

Well, let's say the first retry attempt fails because the network is still down. The `retry` in `onDetached` hasn't returned yet because it supports multiple attempts, but this still triggers another connection `disconnect` event to be fired (why exactly, I'm not sure...it seems that if an attempt at a connection was made and fails, disconnect can still fire). This time the error is slightly different (instead of ECONNRESET, it's something indicating that we can't even resolve the address) and is considered non-retryable.

Now without this check we're doing, this error would surface in the user's error handler. However, we know that the previous `onDetached` is still doing work and should have the final say on whether the connection can come back up or not, so we just log that onDetached was called while one is still active and return early so the 1st onDetached can finish doing what it's doing.

Reply:
If you can add some form of this as a comment in there that would be helpful, i think. As you said it's a semi-unique set of interactions that gets you there.

Reply:
Comments added!

Reply:
Future generations of potentially confused programmers thank you.